src/main/java/org/opensearch/security/filter/SecurityRestFilter.java

```diff
@@ -145,7 +151,56 @@ private boolean userIsSuperAdmin(User user, AdminDNs adminDNs) {
         return user != null && adminDNs.isAdmin(user);
     }
 
-    private boolean checkAndAuthenticateRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception {
+    private boolean authorizeRequest(RestHandler original, RestRequest request, RestChannel channel, User user) {
+
+        List<RestHandler.Route> restRoutes = original.routes();
+        Optional<RestHandler.Route> handler = restRoutes.stream()
+            .filter(rh -> rh.getMethod().equals(request.method()))
+            .filter(rh -> restPathMatches(request.path(), rh.getPath()))
+            .findFirst();
+        if (handler.isPresent() && handler.get() instanceof NamedRoute) {
+            PrivilegesEvaluatorResponse pres = new PrivilegesEvaluatorResponse();
+            NamedRoute route = ((NamedRoute) handler.get());
+            // if actionNames are present evaluate those first
+            Set<String> actionNames = route.actionNames();
+            if (actionNames != null && !actionNames.isEmpty()) {
+                pres = evaluator.evaluate(user, actionNames);
+            }
+
+            // now if pres.allowed is still false check for the NamedRoute name as a permission
+            if (!pres.isAllowed()) {
+                String action = route.name();
+                pres = evaluator.evaluate(user, Set.of(action));
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debug(pres.toString());
+            }
+            if (pres.isAllowed()) {
+                // TODO make sure this is audit logged
+                log.debug("Request has been granted");
+                // auditLog.logGrantedPrivileges(action, request, task);
+            } else {
+                // auditLog.logMissingPrivileges(action, request, task);
+                String err;
+                if (!pres.getMissingSecurityRoles().isEmpty()) {
+                    err = String.format("No mapping for %s on roles %s", user, pres.getMissingSecurityRoles());
+                } else {
+                    err = String.format("no permissions for %s and %s", pres.getMissingPrivileges(), user);
+                }
+                log.debug(err);
+                // TODO Figure out why ext hangs intermittently after single unauthorized request
+                channel.sendResponse(new BytesRestResponse(RestStatus.UNAUTHORIZED, err));
+                return false;
+            }
+        }
+
+        // if handler is not an instance of NamedRoute then we pass through to eval at Transport Layer.
+        // TODO Change this to false once all plugins have migrated to use NamedRoutes
```

Comment:
We will migrate all plugins to use NamedRoutes?  IMO I think we should accept that we will have standard routes used by some plugins and core forever.